feat(test): creating local cluster for e2e

onuryilmaz opened the pull request:
### Creating a local cluster:
```
$ make e2e-local-cluster-create
cd test/e2e/local-cluster && go run .
I0629 13:22:52.739416  580350 create_cluster.go:63] "configuration loaded" kindClusterName="greenhouse-e2e" dockerImageRepository="greenhouse" dockerImageTag="e2e-latest" dockerImageBuildSkip=false kubeconfigName="e2e.kubeconfig" kubeconfigInternalName="e2e.internal.kubeconfig"
I0629 13:23:08.063941  580350 create_cluster.go:80] "cluster created successfully"
I0629 13:23:08.332099  580350 create_cluster.go:108] "kubeconfig exported successfully"
I0629 13:23:09.076845  580350 create_cluster.go:118] "Docker image built successfully"
I0629 13:23:16.396176  580350 create_cluster.go:127] "Docker image loaded to the cluster successfully"
I0629 13:23:39.179051  580350 create_cluster.go:135] "Greenhouse manager is deployed successfully"
```

uwe-mayer commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,103 @@
+	interval = time.Second
+)
+
+var _ = BeforeSuite(func() {
```
I think most of this is covered in
https://github.com/cloudoperators/greenhouse/blob/main/pkg/test/env.go#L98
Which we should uniformely use to boilerplate setup our test-suites.
env-test will also automatically populate the `config` and the `client` it's initalized with when `USE_EXISTING_CLUSTER=true`

So we call `test.TestBeforeSuite` as in https://github.com/cloudoperators/greenhouse/blob/a0a3d4f873f34c0adda7e0d9b009396f69a96f07/test/e2e/suite_test.go#L33

onuryilmaz replied:
outdated, resolving

uwe-mayer commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,103 @@
+	By("Cluster version compatibility")
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	Expect(err).NotTo(HaveOccurred())
+	version, err := discoveryClient.ServerVersion()
```
We might add this assertion (get k8s version) somewhere in the test.BeforeSuite, e.g.:
https://github.com/cloudoperators/greenhouse/blob/main/pkg/test/env.go#L169

onuryilmaz replied:
resolving, outdated

uwe-mayer commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,103 @@
+})
+
+var _ = AfterSuite(func() {
+	By("tearing down the test environment")
```
We should use 

https://github.com/cloudoperators/greenhouse/blob/main/pkg/test/env.go#L173

see exemplarily:

https://github.com/cloudoperators/greenhouse/blob/a0a3d4f873f34c0adda7e0d9b009396f69a96f07/test/e2e/suite_test.go#L38

onuryilmaz replied:
outdated, resolving

uwe-mayer commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,103 @@
+	By("tearing down the test environment")
+})
+
+func createKubeconfigFileForRestConfig(restConfig rest.Config) ([]byte, error) {
```
This hopefully is also already covered by test.BeforeSuite as in:
https://github.com/cloudoperators/greenhouse/blob/0e3dd6fd45aefca4e9e0db3809221a1d67a306ba/pkg/test/env.go#L241-L257

onuryilmaz replied:
outdated, resolving

uwe-mayer reviewed (changes requested):
Nice work!

Take a look at what already exists in `/pkg/test` namely `env.go`. This is what we need to use and leverage to be able to easily switch between contexts:
- env-test
- local kind
- remote greenhouse cluster
It provides a look of tooling that we do not need to duplicate:
- providing a client, config and kubeconfig when testing for example

IvoGoman commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/cloudoperators/greenhouse
 
-go 1.21
+go 1.22.3
```
go1.22 has been merged. I am not the biggest fan of the change that _go.mod_ specifies the whole sematic version of go. This causes problems as soon as the patch version is off by 1. Now the go upgrades are mandated by the 3rd Party Library we use... but there does not really seem to be a way around it.

onuryilmaz replied:
```
$ go mod tidy
go: sigs.k8s.io/e2e-framework@v0.4.0 requires go >= 1.22.3; switching to go1.22.4
go: downloading go1.22.4 (linux/amd64)
```

IvoGoman commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+		kubeconfigInternalName = os.Getenv(kubeconfigInternalNameEnvVar)
+	}
+
+	klog.Info("========== CONFIGURATION ==========")
```
klog is good for now and also compatible with the logr interface. Unification of logger dependencies tracked here #82

onuryilmaz replied:
klog to controller-runtime log eaeedc93cad9cc68c0e8d8627c337c7c6735ce46

IvoGoman commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+		klog.Fatal("Error during file creation:", err)
+	}
+	args := []string{"export", "kubeconfig", "--name", kindClusterName, "--kubeconfig", f.Name()}
+	cmd := exec.Command("kind", args...)
```
Looking at the e2e-framework it seems that os/exec is the way:
https://github.com/kubernetes-sigs/e2e-framework/blob/e2b9f41a232ccc96fdbdafa8fdfacfd953576199/support/kind/kind.go#L153

onuryilmaz replied:
e2e-framework also installs the binary and uses it here while creating the cluster:
https://github.com/kubernetes-sigs/e2e-framework/blob/main/support/kind/kind.go#L155C14-L155C31
https://github.com/kubernetes-sigs/e2e-framework/blob/main/support/kind/kind.go#L253
So we can assume that the binary is available in the local system in this code

IvoGoman commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+	klog.Infof("Greenhouse manager is deployed successfully")
+
+	// Deploy test org
+	err = deployTestOrganization(context.TODO(), k8sClient)
```
Made a proposal for creating K8s resources, onboarding clusters during tests here https://github.com/cloudoperators/greenhouse/pull/237/files#diff-37c27d130341f38233e22a0d737864d93be99a77baf40562d65b5857a8b2a2af
It's currently limited to the tests of TeamRBAC but should then be adapted while touching the other tests.

onuryilmaz replied:
`deployTestOrganization` is removed

IvoGoman reviewed (commented):
I like how this looks now.

onuryilmaz commented on go.mod:
```diff
@@ -36,14 +36,15 @@ require (
-	k8s.io/api v0.29.5
+	k8s.io/api v0.30.1
 	k8s.io/apiextensions-apiserver v0.29.5
-	k8s.io/apimachinery v0.29.5
+	k8s.io/apimachinery v0.30.1
 	k8s.io/cli-runtime v0.29.5
-	k8s.io/client-go v0.29.5
+	k8s.io/client-go v0.30.1
```
reverted back to 1.29.x

onuryilmaz commented on Makefile:
```diff
@@ -156,3 +156,18 @@ golint: $(GOLINT)
+run-e2e-local-cluster: e2e-local-cluster-create
+	export KUBECONFIG=$(shell pwd)/test/e2e/local-cluster/e2e.kubeconfig && \
+	export TEST_E2E_KUBECONFIG_INTERNAL_DOCKER_NETWORK=$(shell pwd)/test/e2e/local-cluster/e2e.internal.kubeconfig && \
+	go test ./test/e2e -v
```
run-e2e added below

onuryilmaz commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+		}
+	}
+
+	if err := wait.For(
```
waiting block is removed

onuryilmaz commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+	klog.Info("Docker image loaded to the cluster successfully")
+
+	// Create Kubernetes client
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigName)
```
this part is removed since `deployTestOrganization` is removed

onuryilmaz commented on pkg/test/env.go:
```diff
@@ -103,6 +105,13 @@ var (
+		useExistingGreenhouseCluster := clientutil.GetEnvOrDefault("TEST_USE_EXISTING_GREENHOUSE_CLUSTER", "false") == "true"
+		if useExistingGreenhouseCluster {
+			// we are making use of https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#pkg-constants to prevent starting a new control plane
+			os.Setenv("USE_EXISTING_CLUSTER", "true")
```
outdated, resolving

onuryilmaz commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,103 @@
+	By("tearing down the test environment")
+})
+
+func createKubeconfigFileForRestConfig(restConfig rest.Config) ([]byte, error) {
```
outdated, resolving

onuryilmaz commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+)
+
+func init() {
+	if os.Getenv(kindClusterNameEnvVar) != "" {
```
updated all of them to use flags (with default values)

onuryilmaz commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,305 @@
+		return err
+	}
+
+	//TODO(onur): make a separate function to edit values
```
IMO, providing values files for `manager` chart should not be part of the cluster creation part. I have tried to use the minimum required to install the chart in the code:
- `dnsDomain` 
- `repository` and `tag` --which can be overridden with the flags 
- `replicas` = 1

onuryilmaz commented:
@IvoGoman  @uwe-mayer  I have updated the files rebasing the main & scoping down to creating local cluster only. It is ready to have another look 👀

uwe-mayer commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,229 @@
+	}
+	controllerManager["image"] = controllerManagerImage
+	controllerManager["replicas"] = "1"
+	chart.Values["controllerManager"] = controllerManager
```
Can we somehow load this config from a values file instead of hardcoding?
E.g.:

```golang
var (
	...

	valuesFileName  = flag.String("valuesFile", "./specs/manager.values.yaml", "path to the values file")
	values          = map[string]interface{}{}
	
    ...
)
....

valuesFile, err := os.ReadFile(*valuesFileName)
if err != nil {
	return ctx, err
}
err = yaml.Unmarshal(valuesFile, &values)
if err != nil {
	return ctx, err
}

....

_, err = installAction.RunWithContext(ctx, chart, values)
```

onuryilmaz replied:
done with cd9b6279c0847fdd58c0d4105dd58d539a8d293c

uwe-mayer commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,229 @@
+	l.Info("Docker image loaded to the cluster successfully")
+
+	// Deploy Greenhouse manager
+	err = installChart(ctx, "./../../../charts/manager", greenhouseControllerManagerRelease, kubeconfig, greenhouseControllerManagerNamespace)
```
I think we also need to install the [DEX (idproxy) CRDs](https://github.com/cloudoperators/greenhouse/tree/main/charts/idproxy/crds), the controller won't come up for me:
```
....
2024-07-03T08:43:10Z	ERROR	controller-runtime.source.EventHandler	if kind is a CRD, it should be installed before calling Start	{"kind": "Connector.dex.coreos.com", "error": "no matches for kind \"Connector\" in version \"dex.coreos.com/v1\""}
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1.1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/source/kind.go:63
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func2
	/go/pkg/mod/k8s.io/apimachinery@v0.29.5/pkg/util/wait/loop.go:87
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext
	/go/pkg/mod/k8s.io/apimachinery@v0.29.5/pkg/util/wait/loop.go:88
k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel
	/go/pkg/mod/k8s.io/apimachinery@v0.29.5/pkg/util/wait/poll.go:33
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/source/kind.go:56
2024-07-03T08:43:10Z	ERROR	Could not wait for Cache to sync	{"controller": "organizationDEX", "controllerGroup": "greenhouse.sap", "controllerKind": "Organization", "error": "failed to wait for organizationDEX caches to sync: timed out waiting for cache to be synced for Kind *api.Connector"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:203
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:208
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/internal/controller/controller.go:234
sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.5/pkg/manager/runnable_group.go:223
2024-07-03T08:43:10Z	INFO	Stopping and waiting for non leader election runnables
2024-07-03T08:43:10Z	INFO	Stopping and waiting for leader election runnables
2024-07-03T08:43:10Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "clusterStatus", "controllerGroup": "greenhouse.sap", "controllerKind": "Cluster"}
2024-07-03T08:43:10Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "pluginHelm", "controllerGroup": "greenhouse.sap", "controllerKind": "Plugin"}
2024-07-03T08:43:10Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "pluginPreset", "controllerGroup": "greenhouse.sap", "controllerKind": "PluginPreset"}
....
```

onuryilmaz replied:
@uwe-mayer  Should we deploy only CRDs or idproxy Helm chart (with scaled to zero)?

onuryilmaz replied:
done with cd9b6279c0847fdd58c0d4105dd58d539a8d293c 

manager pod starts as expected:

```
kubectl logs greenhouse-controller-manager-bb9756c78-947xx
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "teamRoleBindingController"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "pluginPreset"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "bootStrap"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "organizationTeamRoleSeeder"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "teamMembershipPropagation"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "pluginHelm"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "clusterDirectAccess"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "clusterStatus"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "organizationDEX"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "teamPropagation"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "clusterHeadscaleAccess"}
2024-07-11T11:32:19Z    INFO    setup   WARN: headscaleApiKey or headscaleApiUrl is not set
2024-07-11T11:32:19Z    INFO    setup   Skipping headscale access reconciler
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "organizationNamespace"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "organizationRBAC"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "organizationServiceProxy"}
2024-07-11T11:32:19Z    INFO    setup   registering controller  {"name": "teamCAM"}
2024-07-11T11:32:19Z    INFO    setup   registering webhook     {"name": "cluster"}
2024-07-11T11:32:19Z    INFO    controller-runtime.builder      Registering a mutating webhook    {"GVK": "greenhouse.sap/v1alpha1, Kind=Cluster", "path": "/mutate-greenhouse-sap-v1alpha1-cluster"}
2024-07-11T11:32:19Z    INFO    controller-runtime.webhook      Registering webhook       {"path": "/mutate-greenhouse-sap-v1alpha1-cluster"}
2024-07-11T11:32:19Z    INFO    controller-runtime.builder      Registering a validating webhook  {"GVK": "greenhouse.sap/v1alpha1, Kind=Cluster", "path": "/validate-greenhouse-sap-v1alpha1-cluster"}
2024-07-11T11:32:19Z    INFO    controller-runtime.webhook      Registering webhook       {"path": "/validate-greenhouse-sap-v1alpha1-cluster"}
2024-07-11T11:32:19Z    INFO    setup   registering webhook     {"name": "pluginDefinition"}
2024-07-11T11:32:19Z    INFO    controller-runtime.builder      Registering a mutating webhook    {"GVK": "greenhouse.sap/v1alpha1, Kind=PluginDefinition", "path": "/mutate-greenhouse-sap-v1alpha1-plugindefinition"}
2024-07-11T11:32:19Z    INFO    controller-runtime.webhook      Registering webhook       {"path": "/mutate-greenhouse-sap-v1alpha1-plugindefinition"}
...
```

uwe-mayer commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,229 @@
+	return nil
+}
+
+func dockerImageBuild(path string, repoAndtag string) error {
```
The image building does not work for me... 
If I build `greenhouse:e2e-latest` manually and run
```
$ go run . --dockerImageBuildSkip true
I0703 10:39:04.543662   71429 create_cluster.go:63] "configuration loaded" kindClusterName="greenhouse-e2e" dockerImageRepository="greenhouse" dockerImageTag="e2e-latest" dockerImageBuildSkip=true kubeconfigName="e2e.kubeconfig" kubeconfigInternalName="e2e.internal.kubeconfig"
I0703 10:40:34.759070   71429 create_cluster.go:80] "cluster created successfully"
I0703 10:40:35.001250   71429 create_cluster.go:108] "kubeconfig exported successfully"
I0703 10:40:39.955923   71429 create_cluster.go:127] "Docker image loaded to the cluster successfully"
I0703 10:41:20.487547   71429 create_cluster.go:135] "Greenhouse manager is deployed successfully"
```
I get a running controller (see remark about missing Dex CRDs).


But without a manually built image the make target fails.

1. Logging sais image was build successfully
2. Logging errors on loading the image to kind cluster
3. I cannot find the image on my machine (grepping for `e2e` in `docker images` is empty)

```
$ docker rmi greenhouse:e2e-latest
Untagged: greenhouse:e2e-latest
Deleted: sha256:a311d6cfdfd36e3e57c70884d2ab98d31c4fd6c9b1aec8c1fe2f984d6958da87

$ make e2e-local-cluster-create
cd test/e2e/local-cluster && go run .
I0703 11:21:21.821216   86111 create_cluster.go:63] "configuration loaded" kindClusterName="greenhouse-e2e" dockerImageRepository="greenhouse" dockerImageTag="e2e-latest" dockerImageBuildSkip=false kubeconfigName="e2e.kubeconfig" kubeconfigInternalName="e2e.internal.kubeconfig"
I0703 11:21:33.310028   86111 create_cluster.go:80] "cluster created successfully"
I0703 11:21:33.545322   86111 create_cluster.go:108] "kubeconfig exported successfully"
I0703 11:21:48.355849   86111 create_cluster.go:118] "Docker image built successfully"
E0703 11:21:48.377069   86111 create_cluster.go:124] "image load" err="kind: load docker-image greenhouse:e2e-latest failed: exit status 1: exit status 1"
exit status 1
make: *** [Makefile:180: e2e-local-cluster-create] Error 1
$ docker images | grep e2e
```

onuryilmaz replied:
I have added another function to extract error messages from Docker build response by 1ad6ac7b8a592a1d454abc3befc271718272c7c8

Tested with the same approach:

```
> docker rmi greenhouse:e2e-latest
Untagged: greenhouse:e2e-latest
Deleted: sha256:25c652679b8980fa778c238f60ca76cd17bed3f000fe0056bc37227a4cf94b52

> make e2e-local-cluster-create
cd test/e2e/local-cluster && go run .
I0709 18:12:49.448102 1291840 create_cluster.go:69] "configuration loaded" kindClusterName="greenhouse-e2e" dockerImageRepository="greenhouse" dockerImageTag="e2e-latest" dockerImageBuildSkip=false kubeconfigName="e2e.kubeconfig" kubeconfigInternalName="e2e.internal.kubeconfig"
I0709 18:12:49.637200 1291840 create_cluster.go:86] "cluster created successfully"
I0709 18:12:49.883694 1291840 create_cluster.go:114] "kubeconfig exported successfully"
I0709 18:12:50.911422 1291840 create_cluster.go:124] "Docker image built successfully"
I0709 18:12:51.042014 1291840 create_cluster.go:133] "Docker image loaded to the cluster successfully"

 > docker image ls | grep e2e-latest
greenhouse                                                         e2e-latest                                 25c652679b89   About a minute ago   253MB

IvoGoman replied:
This was resolved by calling the **docker** cli directly

onuryilmaz commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,229 @@
+		Dockerfile:     "Dockerfile",
+		Tags:           []string{repoAndtag},
+		Version:        types.BuilderBuildKit,
+		Platform:       "linux/amd64",
```
flag is added by 67a90cbb3f03b1849f4090899017373ff1af2f9d

onuryilmaz commented:
```
⎈ /g/s/g/c/greenhouse > docker rmi greenhouse:e2e-latest
Untagged: greenhouse:e2e-latest
Deleted: sha256:0d48f04c01d1d4fec4e2c16dada00c968101b4ac5d22a8bce19e7d62409f4c71

⎈ /g/s/g/c/greenhouse > kind delete clusters greenhouse-e2e
Deleted nodes: ["greenhouse-e2e-control-plane"]
Deleted clusters: ["greenhouse-e2e"]

⎈ /g/s/g/c/greenhouse > make e2e-local-cluster-create
cd test/e2e/local-cluster && go run .
I0711 11:45:22.514215 2340085 create_cluster.go:82] "configuration loaded" kindClusterName="greenhouse-e2e" dockerImageRepository="greenhouse" dockerImageTag="e2e-latest" dockerImageBuildSkip=false kubeconfigName="e2e.kubeconfig" kubeconfigInternalName="e2e.internal.kubeconfig"
I0711 11:45:39.319226 2340085 create_cluster.go:99] "cluster created successfully"
I0711 11:45:39.555364 2340085 create_cluster.go:127] "kubeconfig exported successfully"
I0711 11:46:23.526305 2340085 create_cluster.go:137] "Docker image built successfully"
I0711 11:46:31.278263 2340085 create_cluster.go:146] "Docker image loaded to the cluster successfully"
I0711 11:46:33.850611 2340085 create_cluster.go:154] "idproxy is deployed successfully"
I0711 11:46:46.652780 2340085 create_cluster.go:162] "Greenhouse manager is deployed successfully"

⎈ /g/s/g/c/greenhouse > export KUBECONFIG=./test/e2e/local-cluster/e2e.kubeconfig
⎈ /g/s/g/c/greenhouse > kubectl get pods -n greenhouse
NAME                                            READY   STATUS      RESTARTS   AGE
greenhouse-controller-manager-bb9756c78-z48lr   1/1     Running     0          23s
greenhouse-kube-webhook-certgen-qf7dq           0/2     Completed   0          23s
⎈ /g/s/g/c/greenhouse > 
```

@uwe-mayer @IvoGoman  fyi

uwe-mayer commented on test/e2e/local-cluster/create_cluster.go:
```diff
@@ -0,0 +1,221 @@
+		l.Error(err, "cluster creation")
+		os.Exit(1)
+	}
+	l.Info("cluster created successfully")
```
Can we also have logging for starting one of the steps.
E.g. "Starting cluster creation"

some steps take long and looking at the logs we do not really know what is happening atm.

Maybe have a look at[ how the `dev-env` spin-up logs](https://github.com/cloudoperators/greenhouse/blob/main/cmd/dev-env/main.go).

Is it possible to include the logs from `kind` or `docker` to get more verbose?

IvoGoman commented on pkg/test/env.go:
```diff
@@ -245,9 +253,7 @@ func StartControlPlane(port string, installCRDs, installWebhooks bool) (*rest.Co
-		kubeConfigLocation := os.Getenv("KUBECONFIG")
-		Expect(kubeConfigLocation).NotTo(BeEmpty(), "the environment variable KUBECONFIG must be set")
-		kubeConfig, err = os.ReadFile(kubeConfigLocation)
+		kubeConfig, err = KubeconfigFromEnvVar("KUBECONFIG")
```
I would propose to rename this environment variable to avoid accidents while running e2e locally but scoped against the wrong environment. With a dedicated environment variable it will be harder to misconfigure the e2e run against the wrong cluster
```suggestion
		kubeConfig, err = KubeconfigFromEnvVar("E2E_KUBECONFIG")
```

IvoGoman commented on Makefile:
```diff
@@ -71,13 +71,21 @@ e2e:
 
+.PHONY: e2e-local-cluster
+e2e-local-cluster: e2e-local-cluster-create  ## Run e2e tests on a local KIND cluster.
+	USE_EXISTING_CLUSTER=true KUBECONFIG=$(shell pwd)/test/e2e/local-cluster/e2e.kubeconfig INTERNAL_KUBECONFIG=$(shell pwd)/test/e2e/local-cluster/e2e.internal.kubeconfig go test ./test/e2e/... -coverprofile cover.out -v
```
```suggestion
	USE_EXISTING_CLUSTER=true E2E_KUBECONFIG=$(shell pwd)/test/e2e/local-cluster/e2e.kubeconfig INTERNAL_KUBECONFIG=$(shell pwd)/test/e2e/local-cluster/e2e.internal.kubeconfig go test ./test/e2e/... -coverprofile cover.out -v
```

IvoGoman commented on Makefile:
```diff
@@ -71,13 +71,21 @@ e2e:
 .PHONY: e2e-remote
-e2e-remote: 
+e2e-remote: ## Run e2e tests against a remote Greenhouse cluster.
 	USE_EXISTING_CLUSTER=true go test ./test/e2e/... -coverprofile cover.out -v
```
Renaming the environment variable is important in this case, as a wrongly scoped `KUBECONFIG` could lead to targeting the wrong cluster.

uwe-mayer commented:
@IvoGoman @onuryilmaz Totally agree on changing the kubeconfig env var to run remote tests against.
See https://github.com/cloudoperators/greenhouse/pull/250/commits/4a8f594a546d973ca2b0b813ac4df03c2caf1ea9
PR is good to go from my side...
